contracts/core/VolOracle.sol

```diff
@@ -0,0 +1,264 @@
+     */
+    function annualizedVol() public view returns (uint256 annualStdev) {
+        return
+            Welford.stdev(accumulator.count, accumulator.m2).mul(
+                annualizationConstant
+            );
+    }
+
+    /**
+     * @notice Returns the TWAP for the entire Uniswap observation period
+     * @return price is the TWAP quoted in quote currency
+     */
+    function twap() public view returns (uint256 price) {
+        (
+            int56 oldestTickCumulative,
+            int56 newestTickCumulative,
+            uint32 duration
+        ) = getTickCumulatives();
+
+        int24 timeWeightedAverageTick =
+            getTimeWeightedAverageTick(
+                oldestTickCumulative,
+                newestTickCumulative,
+                duration
+            );
+
+        // Get the price of a unit of asset
+        // For ETH, it would be 1 ether (10**18)
+        uint128 quoteAmount = uint128(1 * 10**baseCurrencyDecimals);
+
+        return
+            OracleLibrary.getQuoteAtTick(
+                timeWeightedAverageTick,
+                quoteAmount,
+                baseCurrency,
+                quoteCurrency
+            );
+    }
+
+    /**
+     * @notice Gets the time weighted average tick
+     * @return timeWeightedAverageTick is the tick which was resolved to be the time-weighted average
+     */
+    function getTimeWeightedAverageTick(
+        int56 olderTickCumulative,
+        int56 newerTickCumulative,
+        uint32 duration
+    ) private pure returns (int24 timeWeightedAverageTick) {
+        int56 tickCumulativesDelta = newerTickCumulative - olderTickCumulative;
+        int24 _timeWeightedAverageTick = int24(tickCumulativesDelta / duration);
+
+        // Always round to negative infinity
+        if (tickCumulativesDelta < 0 && (tickCumulativesDelta % duration != 0))
+            _timeWeightedAverageTick--;
+
+        return _timeWeightedAverageTick;
+    }
+
+    /**
+     * @notice Gets the tick cumulatives which is the tick * seconds
+     * @return oldestTickCumulative is the tick cumulative at last index of the observations array
+     * @return newestTickCumulative is the tick cumulative at the first index of the observations array
+     * @return duration is the TWAP duration determined by the difference between newest-oldest
+     */
+    function getTickCumulatives()
+        private
+        view
+        returns (
+            int56 oldestTickCumulative,
+            int56 newestTickCumulative,
+            uint32 duration
+        )
+    {
+        IUniswapV3Pool uniPool = IUniswapV3Pool(pool);
+
+        (, , uint16 newestIndex, uint16 observationCardinality, , , ) =
+            uniPool.slot0();
+
+        // Get the latest observation
+        (uint32 newestTimestamp, int56 _newestTickCumulative, , ) =
+            uniPool.observations(newestIndex);
+
+        // Get the oldest observation
+        uint256 oldestIndex = (newestIndex + 1) % observationCardinality;
```

Comment:
just confirming, with the observations array to get first index you are doing new index + 1 modulo the number of observations being store, which will wrap you around the array to oldest index?

Reply:
That's the idea yep. They have a fixed-size array of 65535 observations, and they have to wrap around the index based on the `observationCardinality.

You can see how it behaves here in the [binary search function](https://github.com/Uniswap/uniswap-v3-core/blob/main/contracts/libraries/Oracle.sol#L160-L161).